Linux firmware update

mgsharm opened the pull request:
**Description of changes:**

This PR updates the linux-firmware package to version 20251111 and adds support for AMD GPU firmware. The update refactors the package to use an explicit allowlist approach for firmware selection and creates a separate subpackage for AMD GPU firmware, giving users flexibility in what firmware they install.

  ## Key Changes

  - **Updated linux-firmware**: From version 20230625 to 20251111
  - **Added AMD GPU support**: New `linux-firmware-amdgpu` subpackage containing all AMD GPU firmware
  - **Refactored firmware selection**: Replaced patch-based exclusion with explicit file listing using allowlist strategy
  - **Improved maintainability**: Added comprehensive documentation explaining the approach

  ## Commits in this PR

  ### Commit 1: `linux-firmware: use explicit file listing in spec`

  Implements the initial explicit file listing infrastructure in the spec file:
  - Adds explicit directory structures for firmware we want to include (bnx2x, cxgb4, i915, intel/ice, qed, tigon)
  - Establishes the exclude-then-include pattern for firmware selection
  - Lays groundwork for the allowlist approach

  ### Commit 2: `linux-firmware: update to 20251111 with explicit file listing`

  Major refactoring and version update:
  - **Version bump**: Updates from 20230625 to 20251111
  - **Removes patch-based approach**: Deletes 10 patch files (0001-0010) that were previously used to exclude unwanted firmware
  - **Implements allowlist strategy**: Uses explicit file listing with exclude patterns instead of patches
  - **Updates installation method**: Changes from `make install-zst` to `./copy-firmware.sh --zstd --ignore-duplicates`
  - **Adds comprehensive documentation**: Explains the allowlist (default-deny) approach and why it's used:
    - New firmware added upstream is automatically excluded (safe default)
    - Package size stays minimal and controlled
    - Only ships exactly what we need (~250 files out of 3000+)
    - Documents why a single wildcard exclude won't work
  - **Explicitly lists needed firmware**: Includes specific firmware for:
    - Network: myri10ge, bnx2x, cxgb4, qed, tg3
    - Graphics: Intel i915
    - Network: Intel ICE

  ### Commit 3: `linux-firmware: add amdgpu subpackage`

  Adds AMD GPU firmware support as a separate installable subpackage:
  - **Creates `linux-firmware-amdgpu` subpackage**: Contains all AMD GPU firmware files
  - **Includes AMD GPU licenses**: LICENSE.amdgpu included in the subpackage
  - **Maintains separation**: Main package excludes amdgpu directory, keeping it lightweight
  - **User flexibility**: Users can install just the main package for minimal firmware, or add the amdgpu subpackage if needed for AMD GPU support

  ## Build Results

  ### Generated RPM Packages

  **Main package:**
  - `bottlerocket-linux-firmware-20251111-1.1763772855.143a9de8.br1.x86_64.rpm`
  - Size: 29MB
  - File count: 222 files

  **AMD GPU subpackage:**
  - `bottlerocket-linux-firmware-amdgpu-20251111-1.1763772855.143a9de8.br1.x86_64.rpm`
  - Size: 35MB
  - File count: 678 files

  ```bash
  # Package sizes
  ❯ ls -lah build/rpms/linux-firmware/bottlerocket-linux-firmware-20251111-1.1763772855.143a9de8.br1.x86_64.rpm
  -rw-r--r--. 1 ec2-user ec2-user 29M Nov 22 01:08 build/rpms/linux-firmware/bottlerocket-linux-firmware-20251111-1.1763772855.143a9de8.br1.x86_64.rpm

  ❯ ls -lah build/rpms/linux-firmware/bottlerocket-linux-firmware-amdgpu-20251111-1.1763772855.143a9de8.br1.x86_64.rpm
  -rw-r--r--. 1 ec2-user ec2-user 35M Nov 22 01:08 build/rpms/linux-firmware/bottlerocket-linux-firmware-amdgpu-20251111-1.1763772855.143a9de8.br1.x86_64.rpm

  # File counts
  ❯ rpm -qlp build/rpms/linux-firmware/bottlerocket-linux-firmware-20251111-1.1763772855.143a9de8.br1.x86_64.rpm | wc -l
  217

  ❯ rpm -qlp build/rpms/linux-firmware/bottlerocket-linux-firmware-amdgpu-20251111-1.1763772855.143a9de8.br1.x86_64.rpm | wc -l
  678
```

  **Testing done:**
  - Built an AMI with these changes and successfully launched it on an AMD GPU instance
  - Build AMI with only `linux-firmware-amdgpu` and verified only required AMD firmware packages are installed and not others by checking the inventory:
  ```bash
bash-5.1# cat /var/lib/bottlerocket/inventory/application.json | grep -i firmware -A 5 -B 2
    },
    {
      "Name": "linux-firmware-amdgpu",
      "Publisher": "bottlerocket-kernel-kit",
      "Version": "20251111",
      "Release": "1.1763755682.74d21fbc.br1",
      "Epoch": "0",
      "InstalledTime": "2025-11-22T01:30:15Z",
--
      "Architecture": "x86_64",
      "Url": "https://www.kernel.org/",
      "Summary": "Firmware for amdgpu drivers"
    },
 ```

```bash
  - Confirmed AMD GPU drivers load and initialize successfully:
bash-5.1# dmesg | grep -E 'amdgpu.*initialized'
[   40.521096] amdgpu 0000:51:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   42.757503] amdgpu 0000:51:00.0: amdgpu: SMU is initialized successfully!
[   43.412235] amdgpu 0000:52:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   45.661186] amdgpu 0000:52:00.0: amdgpu: SMU is initialized successfully!
[   46.340865] amdgpu 0000:62:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   48.612670] amdgpu 0000:62:00.0: amdgpu: SMU is initialized successfully!
[   49.317948] amdgpu 0000:63:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   51.593905] amdgpu 0000:63:00.0: amdgpu: SMU is initialized successfully!
[   52.332333] amdgpu 0000:73:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   54.621974] amdgpu 0000:73:00.0: amdgpu: SMU is initialized successfully!
[   55.394387] amdgpu 0000:74:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   57.703321] amdgpu 0000:74:00.0: amdgpu: SMU is initialized successfully!
[   58.496973] amdgpu 0000:84:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   60.814143] amdgpu 0000:84:00.0: amdgpu: SMU is initialized successfully!
[   61.634575] amdgpu 0000:85:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
[   63.959225] amdgpu 0000:85:00.0: amdgpu: SMU is initialized successfully!
  ```  
- Validated that all 674 AMD GPU firmware files required by the kernel modules are present in the package
- Built an AMI to include all `linux-firmware` and varified all packages were available by looking at inventory:
```bash
bash-5.1# cat /var/lib/bottlerocket/inventory/application.json | grep -i linux-firmware -A 10 -B 1
    {
      "Name": "linux-firmware",
      "Publisher": "bottlerocket-kernel-kit",
      "Version": "20251111",
      "Release": "1.1763620283.c4a27102.br1",
      "Epoch": "0",
      "InstalledTime": "2025-11-21T08:42:36Z",
      "ApplicationType": "Unspecified",
      "Architecture": "x86_64",
      "Url": "https://www.kernel.org/",
      "Summary": "Firmware files used by the Linux kernel"
    },
```

yeazelm commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -8,7 +8,7 @@
 
 Name: %{_cross_os}linux-firmware
-Version: 20230625
+Version: 20251021
```
There is a newer version, since we are updating, can we get `linux-firmware-20251111`?

mgsharm replied:
Updated linux-firmware: From version 20230625 to 20251111

yeazelm commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -27,20 +27,53 @@ Source0: https://www.kernel.org/pub/linux/kernel/firmware/linux-firmware-%{versi
-Patch0010: 0010-linux-firmware-amd-ucode-Remove-amd-microcode.patch
+# Base package pulls in both firmware subpackages unconditionally
+Requires: %{name}-amdgpu
+Requires: %{name}-other
```
Let's just make this the base package, with just one additional sub package that is `%{name}-amdgpu`

yeazelm commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -27,20 +27,53 @@ Source0: https://www.kernel.org/pub/linux/kernel/firmware/linux-firmware-%{versi
 %description
 %{summary}.
 
+%package licenses
```
Why would we break the licenses out in a separate package? We should have the licenses correspond to the package they are installed with.

mgsharm replied:
Refactored.

yeazelm commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +86,26 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+%files other
 %dir %{fwdir}
+%exclude %{fwdir}/amdgpu
 %{fwdir}/*
```
Can we confirm what the size is of this package? We run the risk of breaking existing users of this package by expanding the size by a huge portion. In my mind there are maybe three packages if we are going to include all of them, the "main" package which is the normal one that only includes the minimal set like we have today, an `-amdgpu` which looks right in this PR, and a `-extra` or `-other` which includes the remaining firmware we haven't included up until this point.

mgsharm committed:
linux-firmware: use explicit file listing in spec

Signed-off-by: Gaurav Sharma <mgsharm@amazon.com>

mgsharm committed:
linux-firmware: update to 20251111 with explicit file listing

Replace patch-based firmware exclusion with explicit allowlist
approach in spec file. Add documentation explaining the strategy.

Signed-off-by: Gaurav Sharma <mgsharm@amazon.com>

mgsharm committed:
linux-firmware: add amdgpu subpackage

Signed-off-by: Gaurav Sharma <mgsharm@amazon.com>

yeazelm commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +50,473 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+%exclude %{fwdir}/i915/bmg_dmc.bin.zst
+%exclude %{fwdir}/i915/mtl_gsc_1.bin.zst
+%exclude %{fwdir}/i915/xe2lpd_dmc.bin.zst
+%exclude %{fwdir}/i915/xe3lpd_*.bin.zst
```
Let's include these

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
 # kernel support through FW_LOADER_COMPRESS (and FW_LOADER_COMPRESS_ZSTD for kernels >=5.19)
-make DESTDIR=%{buildroot}/ FIRMWAREDIR=%{fwdir} install-zst
+install -d %{buildroot}/%{fwdir}
+./copy-firmware.sh --zstd --ignore-duplicates %{buildroot}/%{fwdir}
```
I see this error in the build logs:
```
WARNING: ignoring option --ignore-duplicates
```

mgsharm replied:
Yes, will remove.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+%{fwdir}/myri10ge_rss_ethp_z8e.dat.zst
+%{fwdir}/phanfw.bin.zst
+
+# Network firmware - Broadcom NetXtreme II (bnx2x)
```
Move the specific directories - `%dir %{fwdir}/bnx2x` - to their corresponding sections.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+
+%files amdgpu
+%license LICENSE.amdgpu
+%dir %{fwdir}
```
This subpackage should not own the same directory as its parent:

```suggestion
%dir %{fwdir}/amdgpu
%{fwdir}/amdgpu/*
```

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+
+# License files
+%license GPL-2
+%license GPL-3
```
By convention, licenses come first in the `%files` list.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+%dir %{fwdir}/intel/ice/ddp-comms
+%dir %{fwdir}/intel/ice/ddp-wireless_edge
+%dir %{fwdir}/qed
+%dir %{fwdir}/tigon
```
Each directory should move to the block listing the files in that directory.

mgsharm replied:
Makes sense.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+# We use an allowlist (default-deny) approach for security and maintainability:
+# - New firmware added upstream is automatically excluded (safe default)
+# - We only ship exactly what we need
+# - Package size stays minimal and controlled
```
By convention, includes come before excludes. It's usually more interesting to see what a package does contain rather than what it doesn't.

mgsharm replied:
Sounds good. Will move includes before.

bcressey commented on packages/linux-firmware/linux-firmware.spec:
```diff
@@ -53,10 +49,471 @@ mkdir -p %{buildroot}/%{fwdir}/updates
+# Structure of excludes below:
+# 1. File pattern excludes (*.fw.zst, *.hex.zst, etc.) - catches loose files
+# 2. Directory excludes - removes entire directories we don't need
+# 3. Specific file pattern excludes - targeted removals for wireless, misc, etc.
```
I would delete these long comments; they won't hold up well to refactoring, since you can't enforce structure through comments.

mgsharm replied:
Makes sense. Will remove them.
